rabbitmq/src/main/java/io/micronaut/rabbitmq/intercept/RabbitMQConsumerAdvice.java

```diff
@@ -372,11 +289,233 @@ private void handleException(RabbitListenerException exception) {
+                 * The registered exception handler of the consumer may handle these cases and it is possible to
+                 * resume message handling by re-registering the consumer on a new channel.
+                 */
+                @Override
+                public void handleShutdownSignal(String consumerTag, ShutdownSignalException sig) {
+                    if (getChannel() instanceof RecoverableChannel && sig.isHardError()) {
+                        // The RabbitMQ client automatic recovery is only triggered by connection errors.
+                        // The consumer will be recovered by the client, so no additional handling here.
+                        LOG.info("The underlying connection was terminated. Automatic recovery attempt is underway for consumer [{}]",
+                                consumerTag);
+                    } else if (channelPool.isTopologyRecoveryEnabled() && getChannel() instanceof RecoverableChannel) {
+                        LOG.info("The channel of this consumer was terminated. Automatic recovery attempt is underway for consumer [{}]",
+                                consumerTag, sig);
+                        synchronized (RecoverableConsumerWrapper.this) {
+                            RecoverableConsumerWrapper.this.consumer = null;
+                            channelPool.returnChannel(getChannel());
+                        }
+                        performConsumerRecovery();
+                    } else {
+                        LOG.error("The channel was closed. Recovery is not enabled. The consumer [{}] will no longer receive messages",
+                                consumerTag, sig);
+                        cancel();
+                    }
+                }
+
+                @Override
+                public void handleDelivery(String consumerTag, Envelope envelope, AMQP.BasicProperties properties, byte[] body)
+                        throws IOException {
+                    // If the broker forces the channel to close, the client may already have prefetched messages in
+                    // memory and will call handleDelivery for these messages, even if they are re-queued by the broker.
+                    // The client will be unable to acknowledge these messages. So it is safe to silently discard
+                    // them, without bothering the callback handler.
+                    if (!getChannel().isOpen()) {
+                        return;
+                    }
+                    if (executorService != null) {
+                        executorService.submit(() -> callbackHandle(envelope, properties, body));
+                    } else {
+                        callbackHandle(envelope, properties, body);
+                    }
+                }
+
+                private void callbackHandle(Envelope envelope, AMQP.BasicProperties properties, byte[] body) {
+                    try {
+                        isHandlingDelivery = true;
```

Comment:
In the case where an executor service is used, depending on the state of the thread pool, it could be exhausted and the action would be queued. In this case `isHandlingDelivery` would be false until the action could be processed at which time the channel may have already been cancelled and returned. `isHandlingDelivery` should probably be moved to before `if (executorService != null)`. In addition instead of a simple boolean it should probably be a counter that would only be considered `false` if the value is 0. This would be required because the `isHandlingDelivery = false` may happen right after `isHandlingDelivery = true` in another thread and thus appearing no work is being done when in fact there is.

Note that is only really applies to cases where autoAck is on